pkg/controller/deactivation/deactivation_controller.go

```diff
@@ -149,10 +165,47 @@ func (r *ReconcileDeactivation) Reconcile(request reconcile.Request) (reconcile.
 	logger.Info("user account time values", "deactivation timeout duration", deactivationTimeout, "provisionedTimestamp", provisionedTimestamp)
 
 	timeSinceProvisioned := time.Since(provisionedTimestamp.Time)
-	if timeSinceProvisioned < deactivationTimeout {
+
+	deactivatingNotificationTimeout := time.Duration((deactivationTimeoutDays-config.Deactivation.DeactivatingNotificationDays)*24) * time.Hour
+
+	if timeSinceProvisioned < deactivatingNotificationTimeout {
+		// It is not yet time to send the deactivating notification so requeue until it will be time to send it
+		requeueAfterTimeToNotify := deactivatingNotificationTimeout - timeSinceProvisioned
+		logger.Info("requeueing request", "RequeueAfter", requeueAfterTimeToNotify,
+			"Expected deactivating notification date/time", time.Now().Add(requeueAfterTimeToNotify).String())
+		return reconcile.Result{RequeueAfter: requeueAfterTimeToNotify}, nil
+	}
+
+	// If the usersignup state hasn't been set to deactivating, then set it now
+	if !states.Deactivating(usersignup) {
+		states.SetDeactivating(usersignup, true)
+
+		logger.Info("setting usersignup state to deactivating")
+		if err := r.client.Update(context.TODO(), usersignup); err != nil {
+			logger.Error(err, "failed to update usersignup")
+			return reconcile.Result{}, err
+		}
+
+		// Requeue so that the deactivation due time can be calculated after the notification has been sent
```

Comment:
The sequence of events here are:

1. Deactivation controller sets the UserSignup state to deactivating if it's not already set (in line 181)
2. Reconciliation is requeued (line 190)
3. UserSignup controller picks up that the deactivating state has been set, and responds by a) creating a pre-deactivating notification for the user, and b) setting the "deactivating notification created" status condition to true.
4. Deactivation controller reconciles again, and if it doesn't find the notification created status condition as expected, requeues again, otherwise:
5. Deactivation controller calculates the amount of time that has passed since the deactivating notification was sent, based on the LastTransitionTime of the condition.  If enough time has now passed, it sets the UserSignup to deactivated.

For step 2) and 4), we could possibly introduce a delay here, perhaps something like 5 seconds might be sufficient.  The reason for the requeue though is correct according to the comment; the reconciliation cannot proceed until the status condition has been set and the time passed since the notification was sent can be calculated.

Reply:
Yeah.. I actually got this flow when continued to read the code below after leaving this comment.
The flow should work but it's quite complicated and not hard to grok from the first glance. Can you please add this comment with detailed flow to the code (I think right here would be a good place)?

Reply:
Comment with detailed workflow has been added.